Fixes for tp_new for C extensions

rudi-c opened the pull request:
Split from https://github.com/dropbox/pyston/pull/1014

In the second commit, I add a special case in `typeCallInner` for `int_cls`, since it's the most common one to show up. It helps performance a bit.
```
# without special case
Comparing to '['baseline']'
              pyston (calibration)                      :   1.01s baseline: 1.02 ( -0.6%)
              pyston django_template3.py                :   1.98s baseline: 1.97 (  0.2%)
              pyston pyxl_bench.py                      :   1.63s baseline: 1.62 (  0.5%)
              pyston sqlalchemy_imperative2.py          :   2.12s baseline: 2.11 (  0.3%)
              pyston django_template3_10x.py            :  11.72s baseline: 11.53 (  1.6%)
              pyston pyxl_bench_10x.py                  :  12.84s baseline: 12.78 (  0.5%)
              pyston sqlalchemy_imperative2_10x.py      :  16.31s baseline: 16.13 (  1.1%)
              pyston (geomean-2ec9)                     :   1.90s baseline: 1.89 (  0.3%)
# with special case
Comparing to '['baseline']'
              pyston (calibration)                      :   1.02s baseline: 1.02 (  0.6%)
              pyston django_template3.py                :   1.99s baseline: 1.97 (  0.7%)
              pyston pyxl_bench.py                      :   1.63s baseline: 1.62 (  0.7%)
              pyston sqlalchemy_imperative2.py          :   2.11s baseline: 2.11 ( -0.1%)
              pyston django_template3_10x.py            :  11.52s baseline: 11.53 ( -0.0%)
              pyston pyxl_bench_10x.py                  :  12.78s baseline: 12.78 (  0.0%)
              pyston sqlalchemy_imperative2_10x.py      :  16.15s baseline: 16.13 (  0.1%)
              pyston (geomean-2ec9)                     :   1.90s baseline: 1.89 (  0.4%)
```

But I don't know how I feel about it, it seems quite hacky for a relatively small performance drop.

kmod commented:
I think if we do this right, there shouldn't be any perf change, since in theory we should be able to get it to the point that these new functions don't get called at all in the benchmarks.

One way to do this would be to have a flag that says "hey even though I have tp_new set, I also have a Python `__new__` available so you should try calling that".  We already have two classes that want this flag (all the exception types, and int), but there are potentially more.  I think the 'is_pyston_class' flag is actually already suitable for this, though maybe a bit hacky.

Another approach is to switch this to the "rewriter aggressiveness backoff" system which is designed for this sort of thing: when we have multiple options for how to try to execute things, but we don't know how many of the fastpaths we'll be able to take.  I think this makes sense, but it's probably overkill and we can keep the change much smaller by checking is_pyston_class.

If you have time, it'd be nice to test putting a stat counter inside that special case, and verify that it gets triggered the same number of times before and after the PR.

kmod commented on src/runtime/str.cpp:
```diff
@@ -1632,11 +1632,11 @@ extern "C" Box* strNonzero(BoxedString* self) {
 }
 
-extern "C" Box* strNew(BoxedClass* cls, Box* obj) {
+template <ExceptionStyle S> Box* strNew(BoxedClass* cls, Box* obj) {
```
Minor, but can you add a `noexcept(S == CAPI)` right before the opening brace?  This will mean that the CAPI version will get a "noexcept" qualifier and is just an extra protection against unhandled exceptions.

rudi-c commented:
I added the noexcept clause.

The performance drop that I initially observed seem to be gone now and the special case in `typeCallInner` hack no longer does anything, so I removed it.

```
Comparing to '['baseline']'
              pyston (calibration)                      :   1.02s baseline: 1.02 ( -0.4%)
              pyston django_template3.py                :   2.08s baseline: 2.08 ( -0.1%)
              pyston pyxl_bench.py                      :   1.80s baseline: 1.82 ( -1.1%)
              pyston sqlalchemy_imperative2.py          :   2.16s baseline: 2.16 ( -0.1%)
              pyston django_template3_10x.py            :  12.33s baseline: 12.27 (  0.5%)
              pyston pyxl_bench_10x.py                  :  13.87s baseline: 13.82 (  0.4%)
              pyston sqlalchemy_imperative2_10x.py      :  16.60s baseline: 16.67 ( -0.4%)
              pyston (geomean-2ec9)                     :   2.01s baseline: 2.02 ( -0.4%)
```

The new case does get called in the benchmarks, setting `is_pyston_class` doesn't change that, but it doesn't improve performance either (it actually makes it worse). But that does lead me to wonder, was there any reason not to set `is_pyston_class` to all of our classes? Performance?

undingen commented on src/runtime/float.cpp:
```diff
@@ -926,6 +926,22 @@ template <ExceptionStyle S> Box* floatNew(BoxedClass* _cls, Box* a) noexcept(S =
+        return NULL;
+
+    if (x == NULL)
+        return floatNew<CAPI>(type, None);
```
I think the arg should be ```boxFloat(0.0)``` not ```None```.
But than I think the ```NULL``` check will never trigger because ```x``` is initialized to ```False``` and marked as optional inside ```PyArg_ParseTupleAndKeywords``` .
I think we could just get rid of the if check in cpython does and just call all the time into ```floatNew<CAPI>(type, x); ```  and similar to cpython make use of the fact that if ```x``` is ```False``` and ```floatNew``` will create a ```0.0``` float for it.

undingen commented:
Hi, sorry for the very late review!
I added a minor comment but decided I don't want to introduce more overhead and even more delays to the whole process and merged it in and created #1115 to fix it. @Daetalus let me know that this fixes some numpy crashes for example ```test_abc.py``` :-) 
Thanks for the patch!

undingen commented:
Oh I forgot to mention: I verified that this does not decrease the performance even though this will cause more tp_new calls...
